cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZookeeperSetOperator.java

```diff
@@ -54,4 +101,226 @@ public static boolean needsRollingUpdate(StatefulSetDiff diff) {
+            log.error("Error while generator Cluster Operator trust store", e);
+        }
+        return trustStore;
+    }
+
+    private SSLSocketFactory socketFactory(String cluster, String namespace) {
+        SSLSocketFactory factory = null;
+        try {
+            CertificateFactory x509 = CertificateFactory.getInstance("X.509");
+            Base64.Decoder decoder = Base64.getDecoder();
+            char[] password = new char[0];
+            SecretOperator secretOperations = new SecretOperator(vertx, client);
+            Secret brokersSecret = secretOperations.get(namespace, ClusterOperator.secretName(cluster));
+            Secret clusterCaKeySecret = secretOperations.get(namespace, cluster + "-cluster-ca");
+            Secret clusterCaCertificateSecret = secretOperations.get(namespace, cluster + "-cluster-ca-cert");
+            CertAndKey coCertKey = Ca.asCertAndKey(brokersSecret, "cluster-operator.key", "cluster-operator.crt");
+            X509Certificate caCertCO = (X509Certificate) x509.generateCertificate(
+                    new ByteArrayInputStream(decoder.decode(clusterCaCertificateSecret.getData().get("ca.crt"))));
+            KeyManagerFactory kmf = KeyManagerFactory.getInstance("SunX509");
+            kmf.init(setupKeyStore(clusterCaKeySecret, x509, password, caCertCO, coCertKey), password);
+            TrustManagerFactory tmf = TrustManagerFactory.getInstance("X509");
+            tmf.init(setupTrustStore(password, caCertCO));
+            SSLContext ctx = SSLContext.getInstance("TLS");
+            ctx.init(kmf.getKeyManagers(), tmf.getTrustManagers(), null);
+            factory = ctx.getSocketFactory();
+        } catch (CertificateException
+                | NoSuchAlgorithmException
+                | UnrecoverableKeyException
+                | KeyStoreException
+                | KeyManagementException e) {
+            log.error("Error while creating Cluster Operator SocketFactory", e);
+        }
+        return factory;
+    }
+
+    private boolean isLeader(Pod pod, SSLSocketFactory factory) {
+        int port = 2181;
+        boolean leader = false;
+        try {
+            if (pod.getStatus() == null) {
+                return false;
+            }
+            String host = pod.getStatus().getPodIP();
+
+            SSLSocket socket = null;
+            Socket plainSocket = factory.createSocket();
+            if (plainSocket instanceof SSLSocket) {
+                // findbugs plugin does not like direct overtyping
+                socket = (SSLSocket) plainSocket;
+            }
+            if (socket == null) {
+                log.error("Could not create socket for getting Zookeeper data");
+                return false;
+            }
+            log.debug("Connecting client to {}:{}", host, port);
+            try {
+                socket.connect(new InetSocketAddress(host, port), 10_000);
```

Comment:
Do we need any changes to the network policies for this? Or do our current Zookeeper rules cover also the CO?

Reply:
I do not think we need any changes. At least all test I did manually work as expected.

Reply:
I am working with 3.10.